### PR TITLE
[plugin.video.catchuptvandmore] 0.2.17

### PR DESCRIPTION
**Pull request by SylvainCecchetto**
### Description

[Version 0.2.17]
[Fix] Live TV - NTV (CA)
[Fix] Live TV - Sport En France (FR)
[Fix] Live TV and Catchup TV - CNews (FR)
[Improvement] Migrate from pickled fav to json data and kill item2dict and item_dict
[Fix] Catch-Up TV - Auvio RTBF (BE)
[Fix] Live TV - CBS News (USA)
[Fix/Improvement] Catch-Up TV - TVA (Canada) - Add search feature
[Fix] Website - Autoplus (FR)
[Fix] Live TV - TV7 Bordeaux (FR)
[Fix] Broken settings
[Fix] Catch-up TV - RTBF (BE)
[Change] Use FranceTV geolocalisation service
[Fix] Catch-up TV - Blaze TV (UK) - Need more work (Cors implementation required ?)
[Fix] Kodi skin - Fix Aura (and others?) widgets
[Fix] Backport typo from master
[Add] Live TV - ABXPLORE / AB3 (BE)
[Add] Catch-up TV - Okoo - France TV (FR)
[Fix] Live TV - NRJ 12 / Cherie 25 (FR)
[Fix] Live TV - TVR (FR)
[Fix] Catch-up TV and Live TV - TL7 (FR)
[Fix] Catch-up TV - Arte (WO)
[Fix] Live TV - Bip TV (FR)
[Remove] Live TV - Canal 10 (FR)
[Fix] Live TV - IDF 1 (FR)
[Fix] Catch-up TV - Noovo (CA)
[Fix] Catch-up TV - TV5 Monde (WO)
[Rename/Change/Add] Refactoring BFM Region (Paris, Lyon, Lille, Grand Littoral) - Add Catch-up TV (BFM Lille, BFM Grand Littoral) (FR) - Anciennement Grand Lille et Grand Littoral TV
[Fix] Catch-up TV - LCP (FR)
[Fix] Catch-up TV - Luxe.TV (FR)
[Rework] Catch-up TV - TV5 Unis (CA)
[Fix] Live TV - NHK World (WO)
[Fix] Live TV - Télé MB (BE)
[Add] Live TV - RTL (FR)
[Add] Catch-up TV - M6 Courses (6play) (FR)
[Add] Website - Fosdem 2020 (BE)

### Checklist:

- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

File: plugin.video.catchuptvandmore/addon.xml [hunk 1]
```diff
@@ -22,7 +22,7 @@
   <extension point="xbmc.addon.metadata">
     <!-- Set reuselanguageinvoker to false while developing the addon, making it only true for the production version -->
     <!-- See https://kodi.wiki/view/Addon.xml#.3Creuselanguageinvoker.3E -->
-    <reuselanguageinvoker>true</reuselanguageinvoker>
+    <!-- <reuselanguageinvoker>true</reuselanguageinvoker> -->
```
Comment (michaelarnauts) [inline plugin.video.catchuptvandmore/addon.xml]: I think you forgot to enable this again.

File: plugin.video.catchuptvandmore/resources/lib/favourites.py [hunk 2]
```diff
@@ -111,145 +172,154 @@ def add_item_to_favourites(plugin, item_dict={}, **kwargs):
...
+        item = menu[k]
+        item_hash = item[1]
+        if item_to_move_id == item_hash:
+            item_to_swap = menu[k + offset]
+            item_to_swap_order = item_to_swap[0]
+            item_to_swap_id = item_to_swap[1]
+            fav_dict[item_to_move_id]['params']['order'] = item_to_swap_order
+            fav_dict[item_to_swap_id]['params']['order'] = item_to_move_order
+            save_fav_dict_in_json(fav_dict)
+            xbmc.executebuiltin('XBMC.Container.Refresh()')
```
Comment (enen92) [inline plugin.video.catchuptvandmore/resources/lib/favourites.py]: same here

File: plugin.video.catchuptvandmore/resources/lib/kodi_utils.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    for word in base_url_l:
+        module_l.append(word)
+
+    module_l.pop()  # Pop the function name (e.g. list_shows)
+    module = '.'.join(module_l)
+
+    return module
+
+
+def import_needed_module():
```
Comment (enen92) [inline plugin.video.catchuptvandmore/resources/lib/kodi_utils.py]: why is this needed?

File: plugin.video.catchuptvandmore/resources/lib/migration_utils.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    The feature (json) appeared in 0.2.17~betaXX
+    All user with version >= 0.2.17 will use item settings in the JSON format
+    Maybe we can remove the migration check on version 0.2.20?
+    """
+
+    # If the json file already exists, the migration has already be done
+    if xbmcvfs.exists(menus_settings_fp):
+        return
+
+    # We directly read the XML setting file (faster than 'xbmcaddon.Addon().getSetting(s)')
```
Comment (enen92) [inline plugin.video.catchuptvandmore/resources/lib/migration_utils.py]: no you don't :) not allowed.

File: plugin.video.catchuptvandmore/resources/lib/addon_utils.py [hunk 5]
```diff
@@ -25,43 +25,72 @@
...
+    # Remote image with complete URL
+    elif 'http' in item_media_path:
+        full_path = item_media_path
 
-    # Build kodi url
-    return urlunsplit(
-        ("plugin", "plugin.video.catchuptvandmore", route_path, query, ""))
+    # Remote image on our images repo
+    else:
+        full_path = 'https://github.com/Catch-up-TV-and-More/images/raw/master/' + item_media_path
```
[the remark this answers is not in the document]
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/addon_utils.py]: Thank you enen92 for your review 👍 
I understand your point of view concerning external resources.
A while ago we made the decision to move all channels logo images in a dedicated GitHub repo because:
* We also use channel logos on our website, so each channel logo is only on one place and both the website and our add-on can access the logo file
* In the past all channel logos were bundle in the add-on. But now we support a lot of channels in our add-on, so with all the images and fanarts the size of our add-on seemed too big for us, this is why we moved images on GitHub.

I will discuss with the co-developer of Catch-Up TV & More in order to find a solution to this 😉

File: plugin.video.catchuptvandmore/resources/lib/kodi_utils.py (hunk 3 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/kodi_utils.py]: This is because we use the CodeQuick framework in our add-on. This one need to register each callback functions to work.
We have a lot of Python files in our add-on (one file for one channel group) and we don't want all the Python files to be loaded on each run. This is why we load on the fly only needed python files for the current Kodi directory.
(This 'hack' will disappear soon because the author of codequick just added a way to manage that point --> https://github.com/willforde/script.module.codequick/commit/131e9023a5778f9c835e802d11f498600428a09c)

File: plugin.video.catchuptvandmore/resources/lib/migration_utils.py (hunk 4 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/migration_utils.py]: Ha ha too bad 😛 
Do you know if there is a function in the API to get all the settings found in `settings.xml` and iterate on it?

File: plugin.video.catchuptvandmore/resources/lib/favourites.py [hunk 6]
```diff
@@ -111,145 +172,154 @@ def add_item_to_favourites(plugin, item_dict={}, **kwargs):
...
-            db[item_hash]['params']['order'] = k
+        menu.append(item)
+    menu = sorted(menu, key=lambda x: x[0])
 
+    for k in range(0, len(menu)):
+        item = menu[k]
+        item_hash = item[1]
+        fav_dict[item_hash]['params']['order'] = k
+    save_fav_dict_in_json(fav_dict)
     xbmc.executebuiltin('XBMC.Container.Refresh()')
```
[the remark this answers is not in the document]
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/favourites.py]: Done

File: plugin.video.catchuptvandmore/resources/lib/favourites.py (hunk 2 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/favourites.py]: Done

File: plugin.video.catchuptvandmore/resources/lib/addon_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (enen92) [reply plugin.video.catchuptvandmore/resources/lib/addon_utils.py]: If the resource addon is included in our repo you can then just link to the image stored in github. You still have full control over the images displayed. Resoruces/images will then be managed as any other addon

File: plugin.video.catchuptvandmore/resources/lib/kodi_utils.py (hunk 3 above)
Comment (enen92) [reply plugin.video.catchuptvandmore/resources/lib/kodi_utils.py]: Please add a #todo comment for removal

File: plugin.video.catchuptvandmore/resources/lib/migration_utils.py (hunk 4 above)
Comment (enen92) [reply plugin.video.catchuptvandmore/resources/lib/migration_utils.py]: I don't know of any way of doing this for addons (only kodi settings). You can create an helper script that reads the settings.xml and dumps the content to some helper file (outside of kodi and don't submit it in this PR). Then in kodi you read that file and not the original one, using the proper api to change the settings.

File: plugin.video.catchuptvandmore/addon.xml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
 <addon id="plugin.video.catchuptvandmore"
        name="Catch-up TV &amp; More"
-       version="0.2.16"
+       version="0.2.17"
```
Comment (enen92) [inline plugin.video.catchuptvandmore/addon.xml]: please add the resource addon as a dependency and make the main addon use the images from the resources addon

Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/addon.xml]: Yes no problem, I will do when the resource add-on will be merged 😉

Comment (enen92) [reply plugin.video.catchuptvandmore/addon.xml]: Resource addon has been aproved so it will be merged when an addon actually depends on it. But since no addon is using it we can't merge it right now :)
So please adjust on this side so we can merge on the other

File: plugin.video.catchuptvandmore/resources/lib/addon_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/addon_utils.py]: Done

File: plugin.video.catchuptvandmore/resources/lib/migration_utils.py (hunk 4 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/migration_utils.py]: Done

File: plugin.video.catchuptvandmore/addon.xml (hunk 7 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/addon.xml]: Done

**Comment by SylvainCecchetto** [comment]
Can you check that the `resource.images.catchuptvandmore` dependency in `addon.xml` is correct, and also that the `resource.images.catchuptvandmore` usage in `addon_utils.py` is also correct?
Thank you very munch!

File: plugin.video.catchuptvandmore/resources/lib/kodi_utils.py (hunk 3 above)
Comment (SylvainCecchetto) [reply plugin.video.catchuptvandmore/resources/lib/kodi_utils.py]: Done

**Comment by enen92** [comment]
I am not familiar with the usage but seems right to me. You can check it by removing your textures folder and restarting kodi. If your icons show up, they come from the resource addon and not from cache

**Comment by SylvainCecchetto** [comment]
> I am not familiar with the usage but seems right to me. You can check it by removing your textures folder and restarting kodi. If your icons show up, they come from the resource addon and not from cache

It works 😉

**Comment by enen92** [comment]
One note: try to submit updates more often, it only takes a bit of time in the case of this addon because each update has more than 200 files to check. If you submit smaller updates we can take care of them much quicker. See the retrospect addon as an example

**Comment by SylvainCecchetto** [comment]
Yes we will try to update more frequently, it's better for you and for users!
Thank you for your time and your review!
